File: app/src/main/java/com/dominikgaller/alpinebooking/booking/core/domain/ParticipantCount.java
package com.dominikgaller.alpinebooking.booking.core.domain;

import com.dominikgaller.alpinebooking.booking.core.domain.exception.InvalidBookingRequestException;

/**
 * Value object representing the number of participants for a booking.
 *
 * <p>Invariant: value must be >= 1. Violation throws {@link InvalidBookingRequestException}.
 *
 * <p>SDD: See {@code documentation/domain/aggregate-tour-booking.spec.md}, section 2a.
 */
public record ParticipantCount(int value) {

    public ParticipantCount {
        if (value < 1) {
            throw new InvalidBookingRequestException(
                    "Participant count must be >= 1, was: " + value);
        }
    }
}

File: src/main/kotlin/com/example/contractmanagement/booking/core/domain/tourbooking/ParticipantCount.kt
package com.example.contractmanagement.booking.core.domain.tourbooking

import com.example.contractmanagement.booking.core.domain.tourbooking.exception.InvalidBookingRequestException

/**
 * How many people a booking is for. Always at least one.
 *
 * Throws a domain exception rather than [IllegalArgumentException] because this value
 * arrives from a command, so the value object is the last guard that is guaranteed to run
 * (`coding-style.definition.md` § 6.2, clause one).
 *
 * SDD: see `documentation/domain/aggregate-tour-booking.spec.md`.
 */
data class ParticipantCount(
    val value: Int,
) {
    init {
        if (value < 1) {
            throw InvalidBookingRequestException("Participant count must be >= 1, was: $value")
        }
    }
}
